src/Constraint/Constraint.php

```diff
@@ -171,6 +171,81 @@ public function versionCompare($a, $b, $operator, $compareBranches = false)
         return version_compare($a, $b, $operator);
     }
 
+    public function compile($otherOperator) {
+        $isBranch = 'dev-' === substr($this->version, 0, 4);
+
+        if (self::OP_NE === $this->operator) {
+            if ($isBranch) {
+                if ($otherOperator === self::OP_EQ) {
+                    return 'false';
+                }
+
+                return 'true';
```

Comment:
This logic looks wrong to me. What if the provided version is the same branch ?

Reply:
This is the current behavior: https://github.com/composer/semver/blob/master/src/Constraint/Constraint.php#L193

in fact this path is requireOperator = `!=` and providedOperators is not `==`. see `if ($otherOperator === self::OP_EQ) {` at line 179

matrix is:
```
!dev-master & !1.0
!dev-master & >1.0
!dev-master & <dev-foo
```